Fix Insecure Default Configuration

Zachary-Squires opened the pull request:
# Description

Changes installOED.sh and docker-compose.yml to automatically generate POSTGRES_PASSWORD and OED_TOKEN_SECRET when OED is launched in production mode. This should remediate an issue where OED_TOKEN_SECRET could be forged to allow a login as an admin. Will also make the postgres database more secure as the password will be much more robust by default. Additionally, installOED.sh now specifically checks if OED_PRODUCTION is no rather than just checking if it is not yes. 

Developed and implemented by:
Zachary Squires - https://github.com/Zachary-Squires
Zachary Bates - https://github.com/Zach-O-Bates

## Type of change

- [ ] Note merging this changes the database configuration.
- [ ] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.org/developer/pr/) ideas
- [x] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.org/developer/cla/) and each author is listed in the Description section.

## Limitations

In order to fully remediate the insecure default configuration we will need to use a separate docker file when launching in development and production, this will be done in a separate PR since it will be intrusive to every OED developer.

Zachary-Squires committed:
Fixes to remediate Issue 2

Zachary-Squires committed:
Update docker-compose.yml

Accidentally pushed with OED_PRODUCTION set to yes

huss commented on docker-compose.yml:
```diff
@@ -13,7 +13,8 @@ services:
             # Custom PGDATA per recommendations from official Docker page
             - PGDATA=/var/lib/postgresql/data/pgdata
-            - POSTGRES_PASSWORD=pleaseChange # default postgres password that should be changed for security.
+            - POSTGRES_PASSWORD=${POSTGRES_PASSWORD:-pleaseChange} # In development password may (and should be) be manually changed. 
```
OED prefers comments above the line so could all the new ones please be changed to do that.

Zachary-Squires replied:
Pretty sure I got them all.

huss commented on docker-compose.yml:
```diff
@@ -111,4 +112,4 @@ services:
         Xvfb :99 -screen 0 1024x768x16"
-        
\ No newline at end of file
+        
```
Could this line be removed since it has spaces at the start and there are two empty lines in a row.

Zachary-Squires replied:
Should be gone I think.

huss commented on src/scripts/installOED.sh:
```diff
@@ -188,11 +188,80 @@ printf "%s\n" "OED install finished"
+			if [ -f ".env" ]; then
+				if grep -q "^OED_TOKEN_SECRET=" .env; then
+					sed -i "s/^OED_TOKEN_SECRET=.*/OED_TOKEN_SECRET=$OED_TOKEN_SECRET/" .env
+				else
```
This comment applies in a general way but put here. I'm thinking about the fact that there are settings in .env and docker-compose.yml that apply to the same case. I think originally it was designed to allow flexibility in how the values were given and to override them. I think the reality is that people don't use this ability. Also, security has become more important so I'm thinking having it in two places is less desirable since people might miss one. Finally, I think OED will be adopting new ways to set values that is more secure and easier so getting to one way now might help for that future effort. First: What do you think about getting rid of one of the ways to set these values? Second: If yes, then should we only use the docker-compose.yml file?

huss replied:
Do you have thoughts on setting it this way? Do you think it should be left for now?

huss replied:
I've decided it could be a separate effort and just opened issue #1569 about doing this.

huss commented on src/scripts/installOED.sh:
```diff
@@ -188,11 +188,80 @@ printf "%s\n" "OED install finished"
+			OED_TOKEN_SECRET=$(openssl rand -hex 32)
+			export OED_TOKEN_SECRET
+
+			printf "Generated OED_TOKEN_SECRET: %s\n" "$OED_TOKEN_SECRET"
```
Curious why here it does not tell them to save but for PG password it does.

Zachary-Squires replied:
Added that, forgot to.

huss commented on src/scripts/installOED.sh:
```diff
@@ -188,11 +188,80 @@ printf "%s\n" "OED install finished"
 		./src/scripts/devstart.sh
-	fi
+	else
+		printf "\nFailure: Invalid or missing enviroment configuration."
```
enviroment misspelled

Zachary-Squires replied:
Fixed

huss commented on src/scripts/installOED.sh:
```diff
@@ -188,11 +188,80 @@ printf "%s\n" "OED install finished"
 if [ "$dostart" == "yes" ]; then
 	if [ "$production" == "yes" ] || [ "$OED_PRODUCTION" == "yes" ]; then
 		printf "%s\n" "Starting OED in production mode"
+		#Checking if the user has set a mail method and left one of the mailing environment variables default, warning if so
```
Could you please add a space after # as done elsewhere in the file.

Zachary-Squires replied:
Fixed

huss commented on src/scripts/installOED.sh:
```diff
@@ -188,11 +188,80 @@ printf "%s\n" "OED install finished"
+		fi
+		#If the user is in production and their token secret has been left default, generating a random one
+		if [ -z "$OED_TOKEN_SECRET" ] || [ "$OED_TOKEN_SECRET" = "?" ]; then
+			printf "\nNo valid OED_TOKEN_SECRET detected. Generating a secure random secret...\n"
```
I think this and other cases should have a banner to highlight the change as was done for the mail warning.

Zachary-Squires replied:
I added the banners on the part where they're told to save or change the value, that seemed like the most important part.

huss commented on src/scripts/installOED.sh:
```diff
@@ -188,11 +188,80 @@ printf "%s\n" "OED install finished"
 if [ "$dostart" == "yes" ]; then
 	if [ "$production" == "yes" ] || [ "$OED_PRODUCTION" == "yes" ]; then
```
This is only one of several places in this script that checks for the type of installation. Different ones are slightly different in the check. For security reasons and to abstract the logic, I am thinking it would be best to decide near the start of the script and set a separate variable to indicates what case is needed. Then that variable can be consistently used for all logic. It also means the check below for an unknown value can happen at the start to quickly stop the install.

Zachary-Squires replied:
I went in and did that, let me know if it's implemented correctly and if I missed manual checks anywhere.

huss commented on src/scripts/installOED.sh:
```diff
@@ -188,11 +188,80 @@ printf "%s\n" "OED install finished"
+		fi
+		#If the user is in production and their postgres password has been left default, generating a random one
+		if [ -z "$POSTGRES_PASSWORD" ] || [ "$POSTGRES_PASSWORD" = "pleaseChange" ]; then
+			printf "\n No valid PostgreSQL password detected. Generating a secure random password...\n"
```
The messages elsewhere don't have a space at the start so I would remove it here.

Zachary-Squires replied:
Fixed

huss commented on src/scripts/installOED.sh:
```diff
@@ -188,11 +188,80 @@ printf "%s\n" "OED install finished"
+				echo "OED_TOKEN_SECRET=$OED_TOKEN_SECRET" > .env
+			fi
+		fi
+		#If the user is in production and their postgres password has been left default, generating a random one
```
I'm not sure this is actually changing the password for Postgres. I'm not an expert but from what I understand, the POSTGRES_PASSWORD environment variable is used when the DB is initialized. This happens at an earlier stage so the original value is used. I'm also unsure if it changes it if done on a later startup. I tried checking with login and it was not clear but seems it might be an issue. We should discuss this.

I don't know how the token is impacted or if that is okay.

Zachary-Squires replied:
Yeah I did some testing and it seems like you're exactly right, for the first time since the database is created before this is run it uses the default password but after that it uses the new one. I'm not sure if there's anything I can do about that but I'm open to ideas.

huss replied:
I thought this had been addressed but my testing did not find it working. I did the following:

- Without OED running, I removed the postgres-data/ directory. This forces complete initialization of the DB on the next startup of OED.
- I edited src/scripts/installOED.sh so it had:
```
if [ "$dostart" == "yes" ]; then
	if [ "$INSTALL_MODE" = "development" ]; then
```
so it used my development environment to force the password initialization to a script generated one.
- Started OED and it said it set the password.
- Shut down OED.
- Edited postgres_data/pg_hba.conf to change trust to password in all local connections to force PG to ask for a password.
- Restart OED.
- Open a shell on the OED Docker database container.
  - Do ``psql -U oed``. It asks for the password. If I give an invalid one it fails. If I use the one output by the first run of the install script then it fails. If I use the default OED PG password then it works. This is why I think it is not changing the password.

Do you have thoughts or ideas? I'm happy to talk with you about this.

Zachary-Squires replied:
I'm not currently randomizing the password for the oed user, just the postgres superuser. I can also do it for oed if you want.

huss replied:
Just noting we talked about this and you are working to make that happen.

huss replied:
Both passwords are now set but, as you told me, you have to restart OED for it to work. I know you are working on this.

huss replied:
Could you provide info on the need to restart OED after the change and if any progress was made to avoid this.

Zachary-Squires replied:
I don't think there's any way to avoid having to restart OED after the change because, to my understanding, the config.js file is initialized at the start and then the config is used in several other files which actually connect to the database. The issue is that the config can't be changed to reflect the new password after it's been sent out, at least as far as I could tell after multiple attempts.

Zachary-Squires replied:
If you have any ideas for me to try obviously I am very open to suggestions.

huss replied:
> If you have any ideas for me to try obviously I am very open to suggestions.

I'm no expert here so take everything with a grain of salt but here are some thoughts:

- The docker-compose.yml file is using ``comand:`` to start the installOED.sh script after the containers are running. Is there any way to run a command before it starts the containers? Then maybe the new password(s) could be set and files edited before the process starts.
- If there is no known way to start OED with the new passwords available then could it automatically be shut down and restarted to reflect the new passwords? This isn't the ideal solution but it would mean that once the process is done then it is running with the new passwords. Trying to limit this to only when the passwords are changed would be desirable.

There may be other options that I'm not thinking of and I'm open to any ideas from anyone. I have not researched how others may have dealt with similar situations.

I hope this helps a little and I'm happy to discuss.

huss replied:
You found another methods and it resolved the issue.

huss reviewed (changes requested):
Thanks to @Zachary-Squires & @Zach-O-Bates for this contribution. I've made some comments to consider. I'm happy to discuss them. I did some testing but not complete as I'm going to wait for any changes to do that.

Zachary-Squires committed:
Update docker-compose.yml to put comments above line

Zachary-Squires committed:
Update docker-compose.yml to remove double empty line with spaces at the end

Zachary-Squires committed:
Implement PR #1554 Suggestions

Added:
1. Banners for the make sure to change this value notifications, also added a notification for the token secret.
2. Fixed spelling error.
3. Added spaces to comments.
4. A variable that checks what the installation type is at the start and is used for all checks.

Zachary-Squires committed:
Removing extra line at the end of docker-compose.yml

huss commented on src/scripts/installOED.sh:
```diff
@@ -186,13 +201,84 @@ printf "%s\n" "OED install finished"
 		npm run start
-	else
-		printf "%s\n" "Starting OED in development mode"
+	elif [ "$INSTALL_MODE" = "development" ]; then
```
Given the checks above, can this now be an else? I'm not sure it is that risky to do development if something went wrong (which it should not).

huss commented on src/scripts/installOED.sh:
```diff
@@ -186,13 +201,84 @@ printf "%s\n" "OED install finished"
+	elif [ "$INSTALL_MODE" = "development" ]; then
+		# Warning the user if they've left their token or postgres password default, we don't randomly generate it in dev mode 
+		if [ -z "$OED_TOKEN_SECRET" ] || [ "$OED_TOKEN_SECRET" = "?" ]; then
+			printf "\n********************************************************************************\n"
```
I'm thinking about the banners here. I think most developers will simply leave the default values as they are generally safe. Given this, do you think there should be banners? If not, how much do you think should be all uppercase?

huss commented on src/scripts/installOED.sh:
```diff
@@ -58,6 +58,21 @@ if [ -f ".env" ]; then
 fi
 
+# Creating a centralized variable to keep track of the type of installation. 
+INSTALL_MODE="invalid"
```
I'm wondering about not initializing to a bad value and changing the second if to be an else on the first one. I think it has the same impact in a clean way in the code. What do you think?

Zachary-Squires committed:
Changes to address feedback on PR #1554

Zachary-Squires committed:
Merge branch 'OpenEnergyDashboard:development' into development

Zachary-Squires committed:
Changes to Address Feedback

Added comments in docker-compose.yml to alert users as to the fact passwords are only drawn from this file once. Added something to the .env file to show the code that the password has already been generated. Changed a redundant if statement to an else.

huss commented on src/scripts/installOED.sh:
```diff
@@ -186,13 +199,74 @@ printf "%s\n" "OED install finished"
+			POSTGRES_PASSWORD=$(openssl rand -base64 12)
+			node ./src/scripts/changePass.js "$POSTGRES_PASSWORD"
+			printf "\n********************************************************************************\n"
+			printf "Generated POSTGRES_PASSWORD: %s\n" "$POSTGRES_PASSWORD"
```
I'm thinking that it might be best to change the "Generated POSTGRES_PASSWORD" to "Generated a secure PostgreSQL password and set PostgreSQL to use it". This will mostly be seen by production sites so the internal name is not important.

Zachary-Squires replied:
I agree, I'll change it.

huss commented on src/scripts/changePass.js:
```diff
@@ -0,0 +1,57 @@
+
+const { Client } = require('pg');
+
+const newPassword = process.argv[2];
```
I'm wondering why this line and the next one are outside the function ``changePassword``. Is that necessary?

Zachary-Squires replied:
I don't think so, pretty sure they were just left out there on accident. I'll move them in.

huss commented on src/scripts/changePostgresPass.js:
```diff
@@ -0,0 +1,57 @@
+
+console.log('Attempting to change postgres user password...');
+
+function escapePassword(password) {
```
I think some comments on what each function is supposed to do and any for difficult to understand blocks of code would be nice.

Zachary-Squires replied:
Will add

huss commented on src/scripts/changePass.js:
```diff
@@ -0,0 +1,57 @@
+		await client.end();
+		
+	} catch (error) {
+		console.error('Error:', error.message);
```
I think it would be nice if the error messages indicated that this was happening during a password change as this may be happening during another process (install).

Zachary-Squires replied:
Will add

huss commented on src/scripts/changePostgresPass.js:
```diff
@@ -0,0 +1,57 @@
+}
+
+// Wait for database to be ready
+setTimeout(async () => {
```
I don't see where this is used in this file and it is not exported.

Zachary-Squires replied:
It's just used to make the script wait before trying to connect, before I had it I was getting ECONNREFUSED errors.

huss commented on src/scripts/changePostgresPass.js:
```diff
@@ -0,0 +1,57 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
I'm thinking the name of this file should include Postgres so it is clear it changes that password.

I'm wondering if there should be a node script command in package.json to call this script so it would be much easier for a site to update the Postgres password if they wanted. I'm not sure if one can easily pass the new password. If not, this isn't going to work.

Zachary-Squires replied:
I will change the file name, as for the second part I thought we talked about it before Christmas break and decided that if someone wanted to change the password after the randomization they could just do it manually. Not a problem if that's changed, I can look into doing that.

huss commented on src/scripts/installOED.sh:
```diff
@@ -186,13 +199,74 @@ printf "%s\n" "OED install finished"
+			printf "\n********************************************************************************\n\n"
+			if grep -q "^POSTGRES_PASSWORD=" .env; then
+				# Don't want to store the password in plain text in the .env file, but need to let the code know the password has been changed
+    			sed -i.bak 's/^POSTGRES_PASSWORD=.*/POSTGRES_PASSWORD=autoSet/' .env 
```
I'm unclear on how the autoSet value is used within the code. Can you help me understand?

Zachary-Squires replied:
When the password has been set automatically set the POSTGRES_PASSWORD variable in the .env file to autoSet so the script can detect that it's already been automatically changed and not re-randomize it. I didn't really want to store the plaintext password in the file or anything. The problem is that without this I can't really access the database to see if a randomization is needed since I need to have the correct password beforehand to access the database from changePass.js

huss reviewed (changes requested):
@Zachary-Squires Thank you for the updated version. A number of comments are now resolved. I added a new comment on a couple of old ones and added a few new ones. Please let me know if you need anything.

Zachary-Squires committed:
Merge branch 'OpenEnergyDashboard:development' into development

Zachary-Squires committed:
Changes to address feedback

Changes to address feedback on 1/19, added/edited comments, changed changePass.js to changePostgresPass.js

Zachary-Squires committed:
Restore executable bit

huss commented on src/scripts/changePostgresPass.js:
```diff
@@ -0,0 +1,58 @@
+		console.error('Error during Postgres password change:', error.message);
+		
+		if (error.message.includes('ECONNREFUSED')) {
+			console.error('Database is not accepting connections yet.');
```
Thanks for clarifying what the timeout was for. In src/scripts/installOED.sh, it also checks for ECONNREFUSED and then does the wait if it happens. It also makes sure it only tries a max number of times (10) and then errors out. I think that might be better than always calling the setTimeout. Thus, I propose a loop in changePassword and wait if there is a connection error with a max number of loops before it reports and error. The wait could then be reduced to 3 seconds as in installOED.sh so it runs faster in most cases. What do you think about that?

huss replied:
What is the status of this?

huss replied:
The location where this happens has moved but I don't see any updates. Could you please comment on what you are thinking.

Zachary-Squires replied:
Yeah I'd agree that your method would probably be better, I kept having an error where postgres wouldn't be ready to take queries by the time the script ran so I just put the timeout in there as a stop gap. I'll change it.

huss reviewed (changes requested):
@Zachary-Squires I'm trying to get all PRs up-to-date so I looked at this one even though you are still working on it. I resolved some comments and left one to think about along with any other open ones.

Zachary-Squires committed:
Changes to address most recent feedback on PR #1554

huss commented on src/scripts/changePostgresPass.js:
```diff
@@ -0,0 +1,96 @@
+// Update .env with PostgreSQL and OED user passwords
+function updateEnvFile(postgresPassword, oedPassword) {
+    let env = '';
+    if (fs.existsSync(ENV_PATH)) {
```
Adding some comments about what the logic is doing in this function would be helpful.

huss commented on src/scripts/changePostgresPass.js:
```diff
@@ -0,0 +1,96 @@
+
+// Generate a secure random password
+function generatePassword() {
+    return crypto.randomBytes(18).toString('base64');
```
The tab indenting has become space indenting. Could it please be changed back. Let me know if you need any help.

huss commented on src/scripts/changePostgresPass.js:
```diff
@@ -0,0 +1,96 @@
+        if (error.message.includes('ECONNREFUSED')) {
+            console.error('Database is not accepting connections yet.');
+        } else if (error.message.includes('password authentication')) {
+            console.error('Authentication failed — default password may already be changed.');
```
Now that it is written to .env, does this still happen regularly or is it a rare event?

I saw that if I tried to change the password twice via manually doing the npm command then it failed this way. I know you told me you were still trying to get it to work after the initial startup and I wonder if this relates to that. What I mean is maybe the process.env.POSTGRES_PASSWORD is not updated on change but I'm very uncertain about this. Whatever it is, it would be nice to track down both the initial setting and future changes. I note it seemed to work fine on the first change of the password via the npm command.

huss replied:
Could you provide info on what you think about this.

Zachary-Squires replied:
The reason you can't change it manually twice via the npm command is the same reason you have to restart OED once when it first gets changed, I can't by any method I've yet seen know what the password is once it's changed so the script can't access the database. This comes back to the same issue I've been having which is that env variables are initialized when the containers are first run and then at least as far as I know can't be reinitialized. I could add a third parameter that lets the user provide the current database password to the script in case they want to run it twice without restarting.

huss commented on docker-compose.yml:
```diff
@@ -13,7 +13,10 @@ services:
             - PGDATA=/var/lib/postgresql/data/pgdata
-            - POSTGRES_PASSWORD=pleaseChange # default postgres password that should be changed for security.
+            # The postgres database will only pull from this value on the first
+            # startup of OED. If a change is desired after this, one must manually
```
Does one still have to use SQL or will the npm command now do it?

huss commented on src/scripts/changePostgresPass.js:
```diff
@@ -0,0 +1,96 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
```
I wondering about the location of this file. All the other files in this directory are scripts of .sh, .bash, etc. Given this, would it make sense and could this be move the src/server/util/?

huss commented on package.json:
```diff
@@ -45,7 +45,8 @@
 		"webData": "node -e 'require(\"./src/server/data/websiteData.js\").insertWebsiteData()'",
-		"addLogMsg": "node -e 'require(\"./src/server/services/addLogMsg.js\").addLogMsgToDB()'"
+		"addLogMsg": "node -e 'require(\"./src/server/services/addLogMsg.js\").addLogMsgToDB()'",
+		"changePostgresPassword": "node ./src/scripts/changePostgresPass.js"
```
Should this directly call changePasswords() as some npm commands do above?

huss commented on src/scripts/changePostgresPass.js:
```diff
@@ -0,0 +1,96 @@
+
+async function changePasswords() {
+    const postgresPassword = process.argv[2] || generatePassword();
+	const oedPassword = process.argv[2] || generatePassword();
```
Not a big deal but should both passwords have to be the same? Maybe use [2] and [3] so can be different?

huss reviewed (changes requested):
Thanks to @Zachary-Squires for addressing many previous comments. I've made a few more to consider and commented on a few older ones. I think this is getting close.

Zachary-Squires committed:
Changes to address comments on PR #1554 from Feb 4th

Zachary-Squires committed:
Merge branch 'OpenEnergyDashboard:development' into development

huss commented on src/scripts/installOED.sh:
```diff
@@ -186,13 +199,63 @@ printf "%s\n" "OED install finished"
+			
+			printf "\n********************************************************************************\n"
+			printf "Generated OED_TOKEN_SECRET: %s\n" "$OED_TOKEN_SECRET"
+			printf "\n Make sure to save or change this value"
```
Could the leading space be removed so msgs have consistent indenting.

Zachary-Squires replied:
Yes I will do that.

huss reviewed (changes requested):
@Zachary-Squires Thank you for the updated version. I'm sorry it took me a while to get back to this. I have gone through all the comments. I added a new comment on any one that is outstanding and these should be the only unresolved ones. Can you please comment on each one so I know what you are thinking. I also added one comment that relates to a previous one but only part of it so I thought this would be easier (new one not resolved but older one is resolved). I also hid the duplicate comments from before to make knowing what is active easier. It may be the case that the new ones get duplicated but I don't yet know. I am waiting on a version that is hopefully the final one to retest everything. Please let me know if anything is not clear or I can help.

Zachary-Squires committed:
Changes to address feedback, no restart on install

Zachary-Squires committed:
Typo

Zachary-Squires committed:
Fix for accidentally setting passwords to the defaults

Zachary-Squires committed:
Merge branch 'development' into development

huss commented on src/server/util/changePostgresPass.js:
```diff
@@ -0,0 +1,158 @@
+
+// Generate a secure random password
+function generatePassword() {
+	return crypto.randomBytes(18).toString('base64');
```
This number (18) is decent but I think 32 (256 bits) is considered very secure. Is there any reason not to use a larger value? I'm not certain it makes a longer password but I doubt anyone is going to memorize the password so it must be stored anyway.

Zachary-Squires replied:
I upped it to 32.

huss commented on src/server/util/changePostgresPass.js:
```diff
@@ -0,0 +1,158 @@
+	}
+
+	// Updating the postgres password if it exists, else appending it instead
+	if (/^POSTGRES_PASSWORD=/m.test(env)) {
```
I think I'm being slow here. test is a RegEx item (I think) from TS. I'm unclear on where "m" is coming from and why not RegEx.test. Could you help me out?

Zachary-Squires replied:
The m just puts it into multiline mode so that the POSTGRES_PASSWORD is checked for in every line until they find it. The test just tells it whether or not POSTGRES_PASSWORD was found in the .env file.

huss commented on src/server/util/changePostgresPass.js:
```diff
@@ -0,0 +1,158 @@
+	return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+// Detect retryable Postgres errors
```
Is it possible to enumerate the errors that are retried in the comment (esp. since some are hex code)? I realize they are well documented online but it would make it easier. Finally, why these specific ones?

Zachary-Squires replied:
I added them, it's these specific ones because I was getting the tuple concurrently updated error and after I looked it up it only happens with those two and is fixed by a retry usually.

huss commented on src/server/util/changePostgresPass.js:
```diff
@@ -0,0 +1,158 @@
+		env = env.trimEnd() + `\nOED_DB_PASSWORD=${oedPassword}\n`;
+	}
+
+	// Writing new passwords to .env or creating it if it doesn't exist yet
```
Maybe add that only allowing the current user to read & write since I think this is for security.

Zachary-Squires replied:
Added.

huss commented on src/server/util/changePostgresPass.js:
```diff
@@ -0,0 +1,158 @@
+			console.log('********************************************************************************');
+			console.log('Generated a secure PostgreSQL and OED password and applied them successfully.');
+			console.log('The passwords have been stored in ".env" for reference.');
+			console.log('If this was run manually, you will need to restart OED for the changes to take effect.');
```
I tested this and if you try to refresh the web page for OED it expires your admin token and then hangs up with console errors trying to load OED. Given this, I think the message should be very clear that OED is currently not working for everyone already using OED until the restart.

Also, could changePostgresPass have another arg that indicates whether this is via the install (so no restart needed) vs. script in package.json (restart needed)? Then the script should know if a restart is needed and highlight it specially in that case. The value could automatically be set by OED so there is no issue (might need to be the first one to avoid missing args from user - maybe a reason to have consts for the index into args so they are easier to change.) What do you think? If this is done, maybe a warning about stopping all current users until restart before it starts and seeking confirmation if doing manually so no one accidentally does this at a bad time.

Zachary-Squires replied:
I changed the message and added the new arg.

huss replied:
I'm resolving given another message about testing this.

huss commented on src/server/util/changePostgresPass.js:
```diff
@@ -0,0 +1,158 @@
+			console.error('Error changing PostgreSQL or OED password:', error.message);
+
+			if (error.message.includes('password authentication')) {
+				console.error('Authentication failed: default password may already be changed.');
```
Maybe add "or password used is incorrect"?

Zachary-Squires replied:
Added.

huss reviewed (changes requested):
First, I want to thank @Zachary-Squires for the ongoing efforts to make this as good as possible. I tried a reasonable level of testing and it seems to work correctly. As a result, I resolved all the previous comments. I did make a few new comments to consider but I'm hopeful they are not anything of significance and finishing this up should not be difficult.

Here are some overall comments to consider:

- I think you did the merge of development to resolve the merge conflict which was nice. However, per the Discord developer channel message that I recently posted and pinned, this breaks OED. To get around this I reset head to the commit before that merge and it worked fine. The other issue needs to be resolved before merging this PR but whatever changes are made hopefully that should not impact this PR.
- For manual reset of password:
  - The short-cut version to start OED without initializing the DB works. (Just for information to keep in the record).
  - It allows any password including very short ones. Should OED require the provided password to be at least some number of characters (say 12)? This should not impact the automatically generated ones which are very long.
- If you stop OED, manually delete the postgres_data/ directory & remove the Docker oed database container then it fails (either with production yes or no). Maybe it is trying to use the new passwords but since it is initializing it starts with the default ones? Whatever the reason, I tried deleting the two Postgres lines in .env (POSTGRES_PASSWORD=... & OED_DB_PASSWORD=...) and starting OED and that did work. This is unusual but sometimes done by developers to fix issues or during testing. However, it is uncommon for developers to change the default passwords and it isn't required in your new system. Unless you have another idea, I say it is fine and we add to the developer docs that this extra step is needed if you changed the passwords.
- What is the status of the limitation listed in the PR description?

Zachary-Squires committed:
Revert commit from OED in docker-compose.yml that makes it crash

Zachary-Squires committed:
Changes to address 4/9 feedback on #1554

huss commented on src/server/util/changePostgresPass.js:
```diff
@@ -0,0 +1,180 @@
+	const isManual = process.argv[4] !== 'install';
+
+	// Warn if manual invocation that all OED users will lose access until restart
+	if (isManual) {
```
I need to test this. I cannot get my version of OED to start. I think it may be due to previous testing where I was manually playing around with files to try to break. Unfortunately, I cannot wipe the DB and restart due to the issue with development. I tried to manually overcome this but it didn't readily work. Thus, I'm noting that testing will be delay until the issue with development is resolved.

I'm also inclined to ask the user if they want to continue and stop if they do not input yes. Would you like to add that or should I? This also probably applies to a similar message below but I need to verify that once I run it.

Zachary-Squires replied:
I went ahead and added it.

huss reviewed (changes requested):
@Zachary-Squires Overall, this looks good. I made one comment about my testing status that suggested a small change. Let me know if you want to do it or not (or have thoughts). Except for being able to fully test this looks except for the comment.

The only other open item was my question about this limitation in the PR description:

> In order to fully remediate the insecure default configuration we will need to use a separate docker file when launching in development and production, this will be done in a separate PR since it will be intrusive to every OED developer.

Could you please comment on the status of this? I want to figure out if there is a plan or an issue needs to be opened.

Please let me know if anything is unclear, have thoughts or need anything.

Zachary-Squires committed:
Adding a confirmation when manually changing the password

huss commented on src/server/util/changePostgresPass.js:
```diff
@@ -0,0 +1,202 @@
+			updateEnvFile(newPostgresPassword, newOedPassword);
+
+			console.log('********************************************************************************');
+			console.log('Generated a secure PostgreSQL and OED password and applied them successfully.');
```
I've managed to get the DB passwords reset so I could test some more. I noticed that this message is always displayed even if the user provided the password. Could it be changed so it says this when generated and, if possible, that it used the user provided ones if not? I do see that it is possible of for only one of the two passwords to be provided to the script so that may make it necessary to do a little tracking.

Zachary-Squires replied:
I looked at it and thought that maybe it just telling the user what passwords were used would accomplish the same purpose, that way you can both verify which method of generation was used and for which password and it won't say "generated" if it was manual. Let me know if the change works.

huss replied:
@Zachary-Squires What you did was fine but I am always a little nervous about dumping a password to a terminal. The person might not notice and leave it so others can see it. I don't think it is a big deal but putting it in a file seems a little better. Of course, if they use the args to set the new password then it is in the history but hopefully they are aware they did this. I'm open to other thoughts.

huss reviewed (changes requested):
@Zachary-Squires Thank you for the update. I believe the PR is down to this:

- I made one new comment to consider.
- There is still one comment (not to a line of code) that needs a response on the status.
- This will have to wait for final testing when the development install issue is resolved. That isn't something this PR needs to do but I am noting.

Please let me know if you have any thoughts or need anything.

Zachary-Squires committed:
Change to message when passwords are changed in changePostgresPass.js

huss commented:
@Zachary-Squires  I'm trying to be sure the status of comments. I think the only older one that I wanted your input is:

> It allows any password including very short ones. Should OED require the provided password to be at least some number of characters (say 12)? This should not impact the automatically generated ones which are very long.

Zachary-Squires committed:
Merge branch 'OpenEnergyDashboard:development' into development

huss committed:
Merge branch 'development' into pr/Zachary-Squires/1554

huss committed:
changed msg for secret token

Tell to see .env as in other places.

huss committed:
12 character passwords

- Enforces 12 character min for all postgres passwords.
- Minor msg formatting update.
- Remove msg with new postgres passwords.

huss committed:
DB startup changes

This is a commit to get a working but not perfect
version saved. It has the following changes:
- Creates shared environment variables for the
web and database containers in docker-compose.yml
and contains the OED DB password.
The goal was to use this so the DB startup could
use that password. Otherwise, on initial startup,
OED would use the default value and then the
OED software would fail when used because it used a
different password from the environment.
Unfortunately, this is not working correctly so the
value is currently duplicated in the DB setup.
- The database build now passes args to send the
desired password. The DB docker file now uses
that value to edit the init.sql file to match.
- Fixes that the install args comments were in the
wrong place.
- The syntax for setting environment variables was
changed to use : (newer way). The ? could not be
used with this value without quoting it. It may be okay
to do others that way but not done.
The leading - was removed as not longer needed/desired. This was done in both docker files.
- Removes the version in docker-compose.yml.
This gets rid of the warning on OED install and is
no longer used.
- A TODO was added since testing found the cypress
UI testing is not working. It crashes after startup. It
is unclear why and the current changes probably
are not the cause.

huss committed:
add comment about changing DB passwords

huss committed:
fix so works better

- This is better but needs more testing/work.
However, it is time to commit.
- Make so knows if production install and
acts accordingly.
- The password check function now returns if
the password should be changed rather than
if it generated a new one so it works in
all cases.
- Put all three environment variables that
can change in .env not just ones changed.
This seemed better since you cannot change
values in the docker config after the initial startup
so now all are in that one file.
- Fixed bug where did not change default
password if it was not changed. Should not
happen once everything works but safer to
make it the first check.
- Added the postgres password to the common
environment variables in docker-compose.yml
since needed by install during web process
to get an password updates.
- Update npm script call to change passwords
so new name and directly calls the function.
This avoids the default running since that
can cause errors if improperly used
and it may not be supported in the future.
- Changed the password change file name to
match script and reflect new usage.
- There are debug statement left in for now.

huss commented:
I've been working on this for several days. I need some more time to get it to a great place given the changes I decided to make. I will push my changes at times to be sure they are not lost but it isn't ready/done until I comment about it.

huss committed:
allow inputted passwords +

- Modify code so user can input either password
when manually changing.
- Fixed up so work in production and with 1 or 2 passwords set.
- Don't change password if input is empty.
No longer need to ask if want to change
if console input.
- Remove unused input function since replaced
with usual OED one.
- Add output if new password same so unchanged.
- Fix issue that script no finish if no passwords
changed and let user know.

huss committed:
few fixes

- Added comment that if either password is
empty in the docker file then install fails.
- Clarified comment that 12 character min
is now enforced for DB passwords.
- Removed check for production of now since
that cannot happen.
- Added TODO comment that check for production
switch does not properly working.

huss committed:
careful read

- Change so empty password always skips
the update.
- Fix a few small items.
- Remove debug statements.

huss commented:
This is to note that this is probably close. I have tried to read carefully and test weird cases. The biggest remaining item is that the common environment variables in the docker-compose.yml file are not doing what is hoped so they are duplicated. It works now but you can cause issues if you don't change both the same way.

@Zachary-Squires I have made non-trivial changes to get it to be how I thought was best (no reflection on you as I changed my mind while looking this over). I know your semester is over but would you be able to look my changes over? Also, line 223 and the following lines in changePostgresPasswords.js have two TODO comments about if it is needed. I was unsure if they were needed to update the values for other places or if it was a hold over from before the later lines set the value involved. Any insight from you would be appreciated.

huss committed:
comment out code

It is believed this is not needed but
left just in case something goes wrong.

huss commented:
> Also, line 223 and the following lines in changePostgresPasswords.js have two TODO comments about if it is needed. I was unsure if they were needed to update the values for other places or if it was a hold over from before the later lines set the value involved. Any insight from you would be appreciated.

I tested commenting it out and it seems to work. Given this, I pushed a commit to do that. Commented out code remains in case something goes wrong along with a comment.

huss commented:
> The biggest remaining item is that the common environment variables in the docker-compose.yml file are not doing what is hoped so they are duplicated. It works now but you can cause issues if you don't change both the same way.

I've created issue #1654 to fix this after this code is merged. I seems more important to get this in to make OED more secure.

huss commented:
Given this helps secure OED, I've decided to merge it even though the original developer has not indicated they review my recent changes. I've tried to be careful but anyone with future thoughts or concerns should let OED know. Since OED is still a while away from a new release, it is hoped that time will allow others to try this out to make sure nothing goes wrong.
